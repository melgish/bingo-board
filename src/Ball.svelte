<script>
  import { createEventDispatcher } from 'svelte';

  // trun when highlighted
  export let lit = false;

  const dispatch = createEventDispatcher();

  $: hot = false;

  function flip() {
    hot = true;
    dispatch('flip');
    setTimeout(() => hot = false, 2000);
  }
</script>

<style>
  div {
    display: grid;
    justify-items: center;
    align-content: center;
    text-align: center;
  }

  span {
    display: grid;
    width: 36pt;
    height: 26pt;
    border-radius: 50px;
    font-size: 18pt;
    font-weight: bold;
  }
  @media screen {
    div {
      background-color: black;
      color: white;
      transition: background-color 2s ease;
      border: 1px solid #300;
    }
    div.hot {
      background-color: #cc0;
    }
    span {
      border-radius: 50%;
      border: 1pt solid currentColor;
      transition: background-color 1s ease;
    }
    .lit span {
      background-color: white;
      color: black;
    }
  }

  @media print {
    div {
      background-color: white;
      color: black;
      border: 1px solid black;
    }
    span {
      border: 1px solid silver;
    }
  }


</style>
<div
  class:hot="{hot}"
  class:lit="{lit}" on:click="{flip}">
  <span><slot/></span>
</div>
