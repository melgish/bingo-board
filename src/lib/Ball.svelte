<script>
  import { createEventDispatcher } from "svelte"
  // true when highlighted
  export let checked = false
  const dispatch = createEventDispatcher()

  $: hot = false

  /**
   * Sends event that switch wants to be flipped
   */
  function flip() {
    hot = true
    dispatch("flip")
    setTimeout(() => (hot = false), 2000)
  }
</script>

<button role="switch" aria-checked={!!checked} class:hot on:click={flip}>
  <div><slot /></div>
</button>

<style>
  [role="switch"] {
    /* strip styling from the button */
    padding: 0;
    margin: 0;
    border: 0;
    border-radius: 0;
    /* change container type */
    display: grid;
    justify-items: center;
    align-content: center;
    text-align: center;
  }
  [role="switch"]:hover,
  [role="switch"]:focus {
    outline: 2px solid red;
  }
  div {
    display: grid;
    width: 40pt;
    height: 36pt;
    border-radius: 50px;
    font-size: 24pt;
    font-weight: bold;
  }
  @media screen {
    [role="switch"] {
      background-color: black;
      color: white;
      transition: background-color 2s ease;
      border: 1px solid #300;
    }
    .hot {
      background-color: #cc0;
    }
    div {
      border-radius: 50%;
      border: 1pt solid currentColor;
      transition: background-color 1s ease;
    }
    [aria-checked="true"] div {
      background-color: white;
      color: black;
    }
  }

  @media print {
    [role="switch"] {
      background-color: white;
      color: black;
      border: 1px solid black;
    }
    div {
      border: 1px solid silver;
    }
  }
</style>
