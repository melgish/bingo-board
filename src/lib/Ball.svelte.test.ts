import {
  afterAll,
  beforeAll,
  beforeEach,
  describe,
  expect,
  it,
  vi,
} from "vitest"
import { act, fireEvent, render, screen } from "@testing-library/svelte"
import Ball from "./Ball.svelte"

// Ball role is a switch
const SWITCH = "switch"
// With checked and unchecked states
const CHECKED = { checked: true }
const UNCHECKED = { checked: false }
// And flip event
const FLIP = "flip"
// With transition highlight
const HOT = "hot"

beforeAll(() => {
  vi.useFakeTimers()
})

afterAll(() => {
  vi.useRealTimers()
})

describe("Ball", () => {
  describe("when not checked", () => {
    it("should have aria-checked=false", () => {
      render(Ball)

      expect(screen.getByRole(SWITCH, UNCHECKED)).toMatchSnapshot()
    })
  })

  describe("when checked", () => {
    it("should have aria-checked=true", () => {
      render(Ball, CHECKED)

      expect(screen.getByRole(SWITCH, CHECKED)).toMatchSnapshot()
    })
  })

  describe("when ball is clicked", () => {
    it("should emit a flip event", async () => {
      const { component } = render(Ball, UNCHECKED)
      const el = screen.getByRole(SWITCH)
      const flipped = vi.fn()

      component.$on(FLIP, flipped)

      await fireEvent.click(el)

      // Make sure class is added / then removed
      expect(el).toHaveClass("hot")
      expect(flipped).toHaveBeenCalled()

      await act(() => vi.runOnlyPendingTimers())

      expect(el).not.toHaveClass(HOT)
    })
  })
})
