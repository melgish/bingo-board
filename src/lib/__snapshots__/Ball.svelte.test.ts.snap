// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Ball > when checked > should have aria-checked=true 1`] = `
<button
  aria-checked="true"
  class="s-H7--1TZSAuX0"
  role="switch"
>
  <div
    class="s-H7--1TZSAuX0"
  />
</button>
`;

exports[`Ball > when not checked > should have aria-checked=false 1`] = `
<button
  aria-checked="false"
  class="s-H7--1TZSAuX0"
  role="switch"
>
  <div
    class="s-H7--1TZSAuX0"
  />
</button>
`;
